sys/net/gnrc/link_layer/lorawan/gnrc_lorawan_mlme.c

```diff
@@ -283,22 +291,134 @@ int _fopts_mlme_link_check_req(lorawan_buffer_t *buf)
         buf->data[buf->index++] = GNRC_LORAWAN_CID_LINK_CHECK_ANS;
     }
-
     return GNRC_LORAWAN_CID_SIZE;
 }
 
-static void _mlme_link_check_ans(gnrc_lorawan_t *mac, uint8_t *p)
+static int _mlme_link_check_ans(gnrc_lorawan_t *mac, uint8_t *p, size_t len, uint8_t index)
 {
+    (void) len;
     mlme_confirm_t mlme_confirm;
 
-    mlme_confirm.link_req.margin = p[1];
-    mlme_confirm.link_req.num_gateways = p[2];
+    assert(p[index] == GNRC_LORAWAN_CID_LINK_CHECK_ANS);
+
+    if (!(mac->mlme.pending_mlme_opts & GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ)) {
+        return 0;
+    }
+
+    mlme_confirm.link_req.margin = p[index + 1];
+    mlme_confirm.link_req.num_gateways = p[index + 2];
+
+    DEBUG("gnrc_lorawan_mlme: LinkCheckAns Margin : %u \n",
+          mlme_confirm.link_req.margin);
+    DEBUG("gnrc_lorawan_mlme: LinkCheckAns GwCnt : %u \n",
+          mlme_confirm.link_req.num_gateways);
 
     mlme_confirm.type = MLME_LINK_CHECK;
     mlme_confirm.status = GNRC_LORAWAN_REQ_STATUS_SUCCESS;
     gnrc_lorawan_mlme_confirm(mac, &mlme_confirm);
 
     mac->mlme.pending_mlme_opts &= ~GNRC_LORAWAN_MLME_OPTS_LINK_CHECK_REQ;
+
+    return GNRC_LORAWAN_FOPT_LINK_CHECK_ANS_SIZE;
+}
+
+static int _mlme_link_adr_req(gnrc_lorawan_t *mac, uint8_t *p, size_t len, uint8_t index)
+{
+    int consumed_bytes = 0;
+    assert(p[index] == GNRC_LORAWAN_CID_LINK_ADR_REQ);
+
+    /* Check whether this is the first block. */
+    if (mac->mlme.adr_flags & 0x8) {
+        /* Indicate error */
+        return 0;
+    }
+
+    /* Mark the first contiguous block */
+    mac->mlme.adr_flags |= 0x8;
+    uint8_t c = index;
+    while (c < len) {
+        if (p[c] == GNRC_LORAWAN_CID_LINK_ADR_REQ) {
+            if ((len - c) >= GNRC_LORAWAN_FOPT_LINK_ADR_REQ_SIZE) {
+                /* There's a LinkADRReq */
+                mac->mlme.adr_req_cnt++;
+                consumed_bytes += GNRC_LORAWAN_FOPT_LINK_ADR_REQ_SIZE;
+            }
+            else {
+                /* The current entry overflow */
+                return 0;
+            }
+        }
+        else {
+            break;
+        }
+        /* Update cursor */
+        c = index + consumed_bytes;
+    }
+
+    int req_index = index + consumed_bytes - GNRC_LORAWAN_FOPT_LINK_ADR_REQ_SIZE;
+
+    uint8_t dr = p[req_index + 1] >> 4;
+    uint16_t tx_power = p[req_index + 1] & 0xF;
+    uint16_t channel_mask = (p[req_index + 3] << 8 ) | p[req_index + 2];
+    uint8_t chmask_ctrl = (p[req_index + 4] & 0x70) >> 4;
+    uint8_t nbtrans = p[req_index + 4] & 0xF;
+
+    int status = 0;
+    gnrc_lorawan_set_uncnf_redundancy(mac, ((nbtrans != 0) ? (uint8_t)(nbtrans - 1) \
+                                             : CONFIG_LORAMAC_DEFAULT_REDUNDANCY));
+
+    if (gnrc_lorawan_set_tx_power(mac, tx_power) >= 0) {
+        status |= 0x4;
+    }
+
+    if (gnrc_lorawan_validate_dr(dr) == true) {
+        mac->last_dr = dr;
+        status |= 0x2;
+    }
+
+    if (chmask_ctrl == 6) {
```

Comment:
What is this number?

Reply:
![ChMaskCntl](https://user-images.githubusercontent.com/5170658/126357687-2a3916cc-f8ae-4c65-adbb-c213b51f4b2e.png)

Reply:
we should use a label (macro) for this value